### ISSUE TITLE
Add adaptive-introgression-related stats?

### ISSUE THREAD

**Issue by grahamgower** [tskit-dev/tskit#1189]
I needed to caclulate some AI-related stats on tree sequences, and have implemented them in the gist linked below. The stats are from Racimo, Marnetto, and Huerta-Sánchez (2017), https://doi.org/10.1093/molbev/msw216. They need some tests, and any other feedback is welcome. I'd be happy to see them integrated into tskit if others think they'll be useful.

https://gist.github.com/grahamgower/2a0ceae81c94e472ed9c509fabadaff7

**Comment by jeromekelleher** [comment]
LGTM - happy to help with the process of implementing them.

**Comment by petrelharp** [comment]
Love it. We could also add the stat from this paper: https://academic.oup.com/mbe/article/36/12/2878/5543093 (even though I think a somewhat different definition would be a bit better).

The hardest part might be working out what to call the new methods. I'm almost tempted to define something like a tree sequence method, `introgression_statistic( , name="xyz")` or something.

I'm also happy to help.

**Comment by jeromekelleher** [comment]
The first step in implementing these @grahamgower is to write down "definitional" versions in Python, see e.g., the [Y1 stat](https://github.com/tskit-dev/tskit/blob/main/python/tests/test_tree_stats.py#L1472). The testing module is a bit of a hairy monster I'm afraid, but hopefully plugging some new stats in there won't be too bad. The C implementation is the last thing to do, once the definitional and Python summary function versions are agreeing. Happy to have a quick call to answer questions here, there is a lot of stuff to take in.

Will these stats have sensible branch/site/node interpretations?

+1 on naming - this is going to be tricky.

**Comment by grahamgower** [comment]
> The first step in implementing these @grahamgower is to write down "definitional" versions in Python, see e.g., the [Y1 stat](https://github.com/tskit-dev/tskit/blob/main/python/tests/test_tree_stats.py#L1472). The testing module is a bit of a hairy monster I'm afraid, but hopefully plugging some new stats in there won't be too bad. The C implementation is the last thing to do, once the definitional and Python summary function versions are agreeing. Happy to have a quick call to answer questions here, there is a lot of stuff to take in.

Ok, I've taken a look a the test code and I think I see what's going on there.

> Will these stats have sensible branch/site/node interpretations?

Umm.. maybe? This stuff makes my head explode, so trying to figure it out is probably a good exercise. Martin et al.'s `\hat{f}_d` might be interesting, as the per-site contribution to the denominator is different depending on whether pop2 or pop3 has the greater allele frequency at the site.

> 
> +1 on naming - this is going to be tricky.

To add another suggestion in to the mix: there's alread `Tajimas_D()`, so we could follow that convention more generally. E.g. `Martin_fd`, `Racimo_U`, etc.

**Comment by jeromekelleher** [comment]
> To add another suggestion in to the mix: there's alread Tajimas_D(), so we could follow that convention more generally. E.g. Martin_fd, Racimo_U, etc.

I was thinking this might be a good approach all right. @petrelharp, thoughts? We might have to add another synonym for the Y stat, but that's OK.

**Comment by petrelharp** [comment]
> To add another suggestion in to the mix: there's alread Tajimas_D(), so we could follow that convention more generally. E.g. Martin_fd, Racimo_U, etc.

Agree - I don't really know what else we'd call them?  (and we probably should have called it e.g. `Pattersons_f4()`.) A point in favor of having an interface like `named_stat(..., method="Racimos_U")` would be that it might make statistics more discoverable? Or, I guess, `ts.stats.Racimos_U( )`?  Maybe we don't need to worry about cluttering up the namespace, though, since most people don't use tab completion in the python shell to find stuff as much as I do?

**Comment by jeromekelleher** [comment]
I agree having a million methods on the TreeSequence object isn't great, but I'm not sure there's much practical options. I think having high-level summary tables like we're doing in msprime at the moment is a good practical way of grouping things docs-wise, although it doesn't solve the problem for tab-completers like you @petrelharp.

The namespacing thing like ``ts.stats.Racimos_U`` is a bit clumsy when you think about the details, but I'm open to more concrete suggestions.

I'm not sure ``named_stat(method=x)`` works though, because we need to allow for different parameters, and we don't want to end up in a matplotlib-like morass of ``kwargs`` passing (although, I do know @grahamgower is a particular fan of this type of API organisation! :wink:)

**Comment by grahamgower** [comment]
Why not a `tskit.stats` submodule? E.g. `tskit.stats.Racimos_U`, with the tree sequence as the first parameter.

**Comment by jeromekelleher** [comment]
> Why not a tskit.stats submodule? E.g. tskit.stats.Racimos_U, with the tree sequence as the first parameter.

Seems kinda backwards to me - it *is* a method of tskit.TreeSequence. It would be annoying and confusing to have to type ``tskit.stats.Racimos_U(ts, ...)`` rather than just ``ts.Racimos_U(...)``.

**Comment by benjeffery** [comment]
Can always add a namespace if we want a bit more organisation: `ts.statistics.racimos_u()`

**Comment by petrelharp** [comment]
If it doesn't seem too bad to you to have them all top-level, then I'm OK with it. I agree, the other options don't seem great.

**Comment by benjeffery** [comment]
Closed for inactivity
